src/lib/integration_test_lib/error_monad.ml

```diff
@@ -0,0 +1,167 @@
+open Core_kernel
+open Async_kernel
+
+module Hard_fail = struct
+  type t =
+    { hard_error: Error.t
+          [@equal
+            fun a b ->
+              String.equal (Error.to_string_hum a) (Error.to_string_hum b)]
+    ; soft_errors: Error.t list
+          [@equal
+            List.equal (fun a b ->
+                String.equal (Error.to_string_hum a) (Error.to_string_hum b) )]
+    }
+  [@@deriving eq, sexp_of, compare]
+end
+
+module Accumulator = struct
+  type 'a t =
+    { computation_result: 'a
+    ; soft_errors: Error.t list
+          [@equal
+            List.equal (fun a b ->
+                String.equal (Error.to_string_hum a) (Error.to_string_hum b) )]
+    }
+  [@@deriving eq, sexp_of, compare]
+end
+
+module ErrorReporting = Monad.Make (struct
+  type 'a t = ('a Accumulator.t, Hard_fail.t) Deferred.Result.t
+
+  let return a =
+    Deferred.return (Ok {Accumulator.computation_result= a; soft_errors= []})
+
+  let bind res ~f =
+    match%bind res with
+    | Ok {Accumulator.computation_result= prev_result; soft_errors= err_list}
+      -> (
+        let execProg = f prev_result in
+        match%bind execProg with
```

Comment:
Because you are just returning in each branch of this match, you can remove the calls to `Deferred.return` on lines 43 and 48 if you change this to a `match%map` instead of a `match%bind`.